### ISSUE TITLE
Why the move away from npm registry?

### ISSUE THREAD

**Issue by JSin** [SheetJS/sheetjs#2667]
I noticed from this ed18acd63d7a5cd31527a36a9af54542cdfcfd30 that you moved away from publishing to the npm registry and recommend people download using tarballs on the CDN. Why did you move away? The npm registry is an extremely common way to download packages.

**Comment by SheetJSDev** [comment]
<img width="542" alt="npm" src="https://user-images.githubusercontent.com/6070939/165421103-c2909aa8-4e58-4452-976a-2cc4dd8e959e.png">

The package on the npm registry is surprisingly popular (one of the top-500 by dependents).  npm invalidated the old publish token and is forcing 2FA on the publishing account.

Due to ongoing legal matters between SheetJS LLC and npm, Inc. (which will not be discussed here), it did not make sense to continue using the public npm registry for distribution.

With [GitHub (the owner of NPM) sunsetting the git.io domain with only 4 days notice](https://github.blog/changelog/2022-04-25-git-io-deprecation/), we are reminded of the ephemerality of the Internet and the inherent risks of relying on platforms.

**Comment by claylevering** [comment]
> we are reminded of the ephemerality of the Internet and the inherent risks of relying on platforms

https://sheetjs.com/careers

> - familiar with the tumult of open source and remote collaboration
> - not prepared to collaborate with the JavaScript and data communities

Top 500 NPM package list, 1.26 million downloads a week, over 3000 dependent packages and not even a **mention** in the README regarding the fact ongoing development for this package won't exist any longer in the single largest JS library platform on the planet (after 8 years).

🤔

**Comment by ljharb** [comment]
Mandatory 2FA should be a noop, as a responsible maintainer would already have it enabled.

npm publish tokens remain exempt from 2FA, so i'm not clear on why that would be an obstacle.

**Comment by jonkoops** [comment]
Yeah, 2FA should be a no brainer. An tokens are indeed except if specified.

**Comment by judehunter** [comment]
This is bizarre

**Comment by lynnntropy** [comment]
Full disclosure, I happened upon this issue by chance and am not a SheetJS user, but this is... really strange.

> The package on the npm registry is surprisingly popular (one of the top-500 by dependents). npm invalidated the old publish token and is forcing 2FA on the publishing account.

What possible justification could you have for taking issue with npm's 2FA requirement for maintainers of popular packages?

> With [GitHub (the owner of NPM) sunsetting the git.io domain with only 4 days notice](https://github.blog/changelog/2022-04-25-git-io-deprecation/), we are reminded of the ephemerality of the Internet and the inherent risks of relying on platforms.

I don't see how this is a reason for silently dropping support for npm. If anything, from the perspective of your users it's an argument for the opposite, because npm is statistically way more likely to exist 5 years from now than your personal CDN. 

Maintainers of OSS projects don't owe people anything, of course, but all I can say is you shouldn't be surprised when people (including your paying customers) look at this whole thing and decide to either fork the project or switch to a competing library not maintained by someone who makes decisions like this.

**Comment by rozzzly** [comment]
I imagine the conversation going somewhere along the lines of:

>  phone rings. _Oh crap it's the CTO! what does he want?!?_

**Hey, I just got an email saying that our _MPM_ account didn't have 2FA enabled. I think that's like really important right? Why didn't you have it enabled?**

> Moment of terror. _My predacessor setup the account... I never thought to check if that 2FA was enabled! But I probably shouldn't say that to the CTO because I definitely should noticed. Um what to say.. uh.. um. come on think! think AHA I got it!_

_Really sir? That is very concerning. They must of uh like um deleted our settings._

**Deleted our settings?! That's outrageous! They can't do that! Those are OUR settings. Just go ahead and only post it on our site from now on.**

Post it on our site? Like a CDN? Well, I'll just reset the security settings no problem.

**No. I doubt these _MPM_ guys will be around a lot longer anyways. You _will_ post it on our site _only_ from now on, am I understood?**

_uh.. yes..._

**Good work, I'm going to go call the lawyers about this.**

> CTO hangs up. _Oh God what did I just do?!_

**Comment by Directory** [comment]
JavaScript hippies back at it again with the tri weekly cdn outages

**Comment by jameshilliard** [comment]
Not sure why one would want to use the sheetjs CDN for npm installs instead of just doing something like this(github based install):
```
npm install SheetJS/sheetjs#semver:^0.18.6
```

**Comment by schw4rzlicht** [comment]
Hilarious.

**Comment by TruffeCendree** [comment]
First, thanks for your amazing work on this library.

Because of security concerns, I prefer relying on npm up-to-date package with proper version management, immutable release binaries and `npm audit` facilities. I'm sure your users would appreciate an update of the well-known npm registry.

If the issue is 2FA related, other popular packages solved the issue. Without you explaining the other reasons, it is hard to understand the withdraw.

Have a nice day.

**Comment by MartinDevillers** [comment]
This has got to be the most bizarre OSS move I've seen since 2016 when [an angry developer unpublished all his 250 npm packages and broke builds all over the planet](https://www.theverge.com/2016/3/24/11300840/how-an-irate-developer-briefly-broke-javascript).

- There are over **two million weekly downloads** on npm pulling in an outdated version of this library. This number is **growing** meaning an **increasing** number of developers are installing **the wrong version**.
- There are **no warnings whatsoever** to let developers know they're using the wrong mechanism to pull in this dependency. The brief mention on the fourth paragraph in the README is **not** sufficient. This needs to be a bright flashing red banner at the very top. The NPM package entry should be updated to reflect this is no longer the latest version and to encourage developers to move over to the CDN. 
- As per [the official documentation](https://docs.sheetjs.com/docs/getting-started/installation/nodejs#vendoring): *"For general stability, "vendoring" modules is the recommended approach:"*. As in, the **recommended approach** is for me to **manually download a tarball** of the latest release and then **add it to my GIT repository**. This feels like going two decades back in time before package managers came into existence. I understand/appreciate this will protect me from supply chain attacks and shoddy internet connections -- but come on. Literally every other package I install will still be coming from the public registry. And if I want to protect myself from supply chain attacks or availability problems of the registry then there are other solutions out there like running a private registry as a personal clone or simply adding node_modules to GIT. 
- To summarize, the motivation for this change is *flimsy*. I understand it's either "I don't want to use MFA" or "I'm in some legal trouble with npm". I don't get it.

Whatever your beef with npm is, please work it out and move on. Don't let it ruin what is otherwise an exceptionally well-designed and well-maintained project.

**Comment by nrutman** [comment]
Just wanted to second the motion for a big update to the README and a warning when installing from the public registry. This is very abnormal and I (like many others, most likely) did not realize I was running on an old version of this package.

**Comment by marracuene** [comment]
Came across this due to this recently-announced vulnerability (our build pipeline is configured to fail if we have deps containing vulns of a certain level):  https://github.com/advisories/GHSA-4r6h-8v6p-xvw6

XLSX CE is a great resource, and free, and indeed it is the maintainer's right to host versions wherever they choose.
However, the stated reason for moving it away from NPM "I don't want to use MFA" does lead us to question whether this is a safe package to continue using - given that it will always be rather vulnerable to supply-chain attacks.    We had implemented some small POC features with it, and were about to add more.

UPDATE 14/06/23: we ended up switching to https://www.npmjs.com/package/export-from-json - a much smaller feature set than sheetjs, but does just what we need, small and no deps.

**Comment by srl295** [comment]
Why the move away from GitHub?

**Comment by YogliB** [comment]
@marracuene how flexible is it?
I had to do some complex Excel manipulations at work and SheetJS was the only package I found that's relatively maintained and can do complicated stuff...

**Comment by marracuene** [comment]
> @marracuene how flexible is it?
> I had to do some complex Excel manipulations at work and SheetJS was the only package I found that's relatively maintained and can do complicated stuff...

@YogliB I suspect it would not handle your use case. In our use case we already have, in-memory, an array of Objects that represent the information to be exported via Excel. All the work to prepare this information has already been done.

The only additional work done at time of export (and which the new package allows us to do), is custom-formatting on specific fields.

**Comment by codeams** [comment]
I know, I know,
nobody should be installing stuff GPT-4 recommends without proper assessment first.

But damn, GPT and any search engine out there straight up recommending to just `npm install xslx` (like you would do in any reasonable, well maintained package).

After finally finding relevant docs (you really have to search for it), reading through a bunch of almost irrelevant demos, still couldn't understand why the npm package hasn't been updated in over a year.

Is it so good that it doesn't have bugs or improvements to make?

Nah, I get it now.

This was a fantastic opportunity to switch from work to my 8pm routine of dinner and watching comedy.

**Comment by srl295** [comment]
@SheetJSDev 

> it did not make sense to continue using the public npm registry for distribution.

Hi, it's been a year since you wrote that. Please consider updating your npm entry to reflect the vulnerability status of the last posted version and to point people in the right direction? I think it would be the right thing to do for the community. Thanks!

**Comment by e965** [comment]
Hey, folks! I made a little tool that allows you to continue using xlsx in your projects.
More details are described here https://github.com/SheetJS/sheetjs/issues/2822#issuecomment-1668303942

I'm writing here as well, so that more people interested in solving the problem will get notified 🙂

**Comment by gierschv** [comment]
Maybe this package should be marked as deprecated on NPM? 🤔
That doesn't make sense to keep it available like that with an outdated and with vulnerabilities in it?

**Comment by srl295** [comment]
I suggested that.

Do the responsible thing, SheetJS!

**Comment by esschul** [comment]
Pretty please with sugar on top, please release on NPM, You'll fix millions of problems. You remember that song from Mariah Carey? "And then a hero comes along, with the strength to carry on". That will be playing in the background as you do it.

**Comment by jakemitchellxyz** [comment]
oof.

**Comment by ryenus** [comment]
> Top 500 NPM package list, 1.26 million downloads a week, over 3000 dependent packages and not even a mention in the README regarding the fact ongoing development for this package won't exist any longer in the single largest JS library platform on the planet (after 8 years).

Is sheetjs sold or will it be?

**Comment by NoNameProvided** [comment]
Since the CE edition is licensed under Apache which allows republishing it with proper attribution I have made a [small automated script](https://github.com/NoNameProvided/xlsx-republish) that periodically checks the CDN and republishes the latest version to NPM if needed. 

The NPM package can be found here https://www.npmjs.com/package/xlsx-republish.

**Comment by marracuene** [comment]
> Since the CE edition is licensed under Apache which allows republishing it with proper attribution I have made a [small automated script](https://github.com/NoNameProvided/xlsx-republish) that periodically checks the CDN and republishes the latest version to NPM if needed.
> 
> The NPM package can be found here https://www.npmjs.com/package/xlsx-republish.

What a great public contribution.  Well done.   We actually moved a way from sheetjs a while ago due to this whole weird soap opera, but you will surely save lots of effort for lots of peeps.

**Comment by jdanyow** [comment]
looks like there's two republished npm packages:
1. [@e965/xlsx](https://www.npmjs.com/package/@e965/xlsx) (most downloads- [see comment above](https://github.com/SheetJS/sheetjs/issues/2667#issuecomment-1668351173))
2. [xlsx-republish](https://www.npmjs.com/package/xlsx-republish) (recently created- [see comment above](https://github.com/SheetJS/sheetjs/issues/2667#issuecomment-2354334791))

**Comment by srl295** [comment]
> They couldn't figure out how to charge money for the freemium version with NPM. Hard pass.

So you are saying that the undisclosed legal challenges with NPM that have been mentioned here, have to do with trying to charge for something that is published? That really puts a very different spin on things.

**Comment by srl295** [comment]
As I noted above, the responsible thing to do would be to mark the versions that are in NPM as deprecated, with a warning.

**Comment by Pirulax** [comment]
@e965 from what I see your repo hasn't been updated lately? Or has the upstream project not change since?

**Comment by e965** [comment]
@Pirulax The repo contains only the code for publishing the package, and it does not need to be updated unless necessary.

**Comment by noren95** [comment]
@e965 is this package https://www.npmjs.com/package/@e965/xlsx official or just an individually-created distribution that supports the CDN releases on the package manager level?

**Comment by diesieben07** [comment]
It would be nice if at least `cdn.sheetjs.com` could be turned into a proper NPM registry, so that tools like `npm outdated` will report an outdated sheetjs version.

**Comment by nickredmark** [comment]
Who else landed here because cloudflare down? 😆
